Realm/RLMSyncUser.mm

```diff
@@ -413,30 +417,44 @@ + (void)_performLogInForUser:(RLMSyncUser *)user
                                                                        requireRefreshToken:YES];
             if (!model) {
                 // Malformed JSON
-                completion(nil, make_auth_error_bad_response(json));
+                NSError *badResponseError = make_auth_error_bad_response(json);
+                dispatch_async(callbackQueue, ^{
+                    completion(nil, badResponseError);
+                });
                 return;
             } else {
                 std::string server_url = authServerURL.absoluteString.UTF8String;
                 SyncUserIdentifier identity{[model.refreshToken.tokenData.identity UTF8String], std::move(server_url)};
                 auto sync_user = SyncManager::shared().get_user(identity , [model.refreshToken.token UTF8String]);
                 if (!sync_user) {
-                    completion(nil, make_auth_error_client_issue());
+                    NSError *authError = make_auth_error_client_issue();
+                    dispatch_async(callbackQueue, ^{
+                        completion(nil, authError);
+                    });
                     return;
                 }
                 sync_user->set_is_admin(model.refreshToken.tokenData.isAdmin);
                 user->_user = sync_user;
-                completion(user, nil);
+                dispatch_async(callbackQueue, ^{
+                    completion(user, nil);
+                });
             }
         } else {
             // Something else went wrong
-            completion(nil, error);
+            dispatch_async(callbackQueue, ^{
+                completion(nil, error);
+            });
         }
     };
     [RLMNetworkClient sendRequestToEndpoint:[RLMSyncAuthEndpoint endpoint]
                                      server:authServerURL
                                        JSON:json
                                     timeout:timeout
-                                 completion:handler];
+                                 completion:^(NSError *error, NSDictionary *dictionary) {
+                                     dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
Should this be using `callbackQueue` rather than the main queue here? Should we have tests verifying that each of the possible callbacks are made on the requested queue?